### DISCUSSION TITLE
Does moor supports fts4?

### DISCUSSION THREAD

**Discussion by PurveshModi** [simolus3/drift#1389] [Q&A]
@simolus3  I want to implement the fts(i.e. Full Text Search) on my flutter application using the fts4 sqlite module. Can you please guide me or provide me any documentation for the same.

Thanks in advance.

**Comment by simolus3** [comment]
Moor does not have builtin support for fts4. Moor supports static analysis for tables and queries with fts5 (in moor files), but there's no such thing for fts4. Why do you want to use fts4 instead of the newer fts5? If it's for compatibility with older sqlite3 versions, you can depend on `sqlite3_flutter_libs` and use a `VmDatabase` to get the latest sqlite3 version.

I can take a look at supporting older fts versions as well, but it's quite some work for (I assume) few users so I'm wondering why you prefer fts4.

**Reply by PurveshModi** [reply]
Okay, thanks for this detailed explanation. I want to use FTS4 because of a compatibility issue, but as you said sqlite3_flutter_libs will take care of it so now I will definitely use FTS5.

For implementing FTS5 using moor, I need quite detailed documentation if you can provide me.
Cause I have some questions in my mind. Can we create a DAO for the FTS table like a normal SQLite table? If yes, then please provide me some steps on how can we do that?

Note: I've created an FTS table using the .moor file.

**Reply by simolus3** [reply]
For a FTS5 table, moor generates the classes you'd expect (a row class, a companion and the table class itself). You can use them to build a DAO for that table supporting inserts and general selects.

For search queries, I'd probably put them into a moor file as well. If you had a file defining the table:

```sql
-- table.moor, included from the main database

CREATE VIRTUAL TABLE documents USING fts5(title, text);
```

Then you can have another moor file defining the queries:

```sql
-- this moor file is only included on the dao, which adds the necessary methods
import 'table.moor';

findDocuments: SELECT * FROM documents WHERE documents MATCH ? ORDER BY rank;
```

**Reply by PurveshModi** [reply]
In this case, do I need to pass the table class in @UseDao annotation?

And Is it mandatory to create another moor file for writing the search queries? Or can we include it in the same moor file? In the above example table.moor file.

**Reply by simolus3** [reply]
> In this case, do I need to pass the table class in @UseDao annotation?

No, that's not necessary. But you'd have to add the moor file to the `include` of the `@UseDao` annotation.

> And Is it mandatory to create another moor file for writing the search queries?

No, it is not mandatory. But if you add the query to the main moor file (defining the table, then moor would create a method for the query on your database class). If you split the moor files, you can make sure that the query is only added to the DAO.

**Reply by PurveshModi** [reply]
@simolus3 I am facing the below error
SqliteException(1): parse error in "id NOT NULL DEFAULT ''"

tables.moor file query:
```sqlite
CREATE VIRTUAL TABLE IF NOT EXISTS donors USING fts5(
  id NOT NULL DEFAULT '',
  lastNameEnglish UNINDEXED NOT NULL DEFAULT '',
  firstNameEnglish UNINDEXED NOT NULL DEFAULT '',
  lastNameHebrew UNINDEXED NOT NULL DEFAULT '',
  firstNameHebrew UNINDEXED NOT NULL DEFAULT '',
  titleHebrew UNINDEXED NOT NULL DEFAULT '',
  titleEnglish UNINDEXED NOT NULL DEFAULT '',
  suffixHebrew UNINDEXED NOT NULL DEFAULT '',
  address1 NOT NULL DEFAULT '',
  address2 NOT NULL DEFAULT '',
  city UNINDEXED NOT NULL DEFAULT '',
  state UNINDEXED NOT NULL DEFAULT '',
  zip UNINDEXED NOT NULL DEFAULT '',
  fatherNameHebrew UNINDEXED NOT NULL DEFAULT '',
  fatherInLawNameHebrew UNINDEXED NOT NULL DEFAULT '',
  parentId UNINDEXED NOT NULL DEFAULT '',
  email UNINDEXED NOT NULL DEFAULT '',
  note UNINDEXED NOT NULL DEFAULT '',
  paternalGrandParentId UNINDEXED NOT NULL DEFAULT '',
  paternalGrandParentId2 UNINDEXED NOT NULL DEFAULT '',
  maternalGrandParentId UNINDEXED NOT NULL DEFAULT '',
  maternalGrandParentId2 UNINDEXED NOT NULL DEFAULT '',
  inactive UNINDEXED NOT NULL DEFAULT '',
  lastUpdated UNINDEXED NOT NULL DEFAULT '',
  accountId UNINDEXED NOT NULL DEFAULT '',
  fullNameEnglish NOT NULL DEFAULT '',
  fullNameHebrew NOT NULL DEFAULT '',
  phone1 NOT NULL DEFAULT '',
  phone2 NOT NULL DEFAULT '',
  phone3 NOT NULL DEFAULT '',
  phone4 NOT NULL DEFAULT '',
  phone5 NOT NULL DEFAULT '',
  phone6 NOT NULL DEFAULT ''
);
```

Generated FTS table query
```dart
  @override
  String get moduleAndArgs =>
      'fts5(id NOT NULL DEFAULT \'\', lastNameEnglish UNINDEXED NOT NULL DEFAULT \'\', firstNameEnglish UNINDEXED NOT NULL DEFAULT \'\', lastNameHebrew UNINDEXED NOT NULL DEFAULT \'\', firstNameHebrew UNINDEXED NOT NULL DEFAULT \'\', titleHebrew UNINDEXED NOT NULL DEFAULT \'\', titleEnglish UNINDEXED NOT NULL DEFAULT \'\', suffixHebrew UNINDEXED NOT NULL DEFAULT \'\', address1 NOT NULL DEFAULT \'\', address2 NOT NULL DEFAULT \'\', city UNINDEXED NOT NULL DEFAULT \'\', state UNINDEXED NOT NULL DEFAULT \'\', zip UNINDEXED NOT NULL DEFAULT \'\', fatherNameHebrew UNINDEXED NOT NULL DEFAULT \'\', fatherInLawNameHebrew UNINDEXED NOT NULL DEFAULT \'\', parentId UNINDEXED NOT NULL DEFAULT \'\', email UNINDEXED NOT NULL DEFAULT \'\', note UNINDEXED NOT NULL DEFAULT \'\', paternalGrandParentId UNINDEXED NOT NULL DEFAULT \'\', paternalGrandParentId2 UNINDEXED NOT NULL DEFAULT \'\', maternalGrandParentId UNINDEXED NOT NULL DEFAULT \'\', maternalGrandParentId2 UNINDEXED NOT NULL DEFAULT \'\', inactive UNINDEXED NOT NULL DEFAULT \'\', lastUpdated UNINDEXED NOT NULL DEFAULT \'\', accountId UNINDEXED NOT NULL DEFAULT \'\', fullNameEnglish NOT NULL DEFAULT \'\', fullNameHebrew NOT NULL DEFAULT \'\', phone1 NOT NULL DEFAULT \'\', phone2 NOT NULL DEFAULT \'\', phone3 NOT NULL DEFAULT \'\', phone4 NOT NULL DEFAULT \'\', phone5 NOT NULL DEFAULT \'\', phone6 NOT NULL DEFAULT \'\')';
```

Can you please help me to figure out the issue?
A couple of hours ago it was working as expected.

**Reply by simolus3** [reply]
>  A couple of hours ago it was working as expected.

That must be because you didn't have things like `NOT NULL` or `DEFAULT` in there an hour ago :D Quoting the [sqlite3 docs](https://www.sqlite.org/fts5.html):

> All type information or constraint specifications must be removed from column definitions. FTS3/4 ignores everything following the column name in a column definition, FTS5 attempts to parse it (and will report an error if it fails to). 

So it looks like all of that `NOT NULL DEFAULT ''` doesn't make a difference or not and can be removed.

**Reply by PurveshModi** [reply]
Thank You for the quick solution.
I really appreciate the support from your side.
